### ISSUE TITLE
Better integration with XML date/time classes

### ISSUE THREAD

**Issue by jodastephen** [ThreeTen/threeten-extra#15]
Provide tools to integrate with the XML classes.

**Comment by raupachz** [comment]
What XML classes are we talking about? ` javax.xml.bind`?

**Comment by jodastephen** [comment]
`XMLGregorianCalendar` and `Duration`.

**Comment by raupachz** [comment]
Ok, what you have in mind? Adapter pattern?

**Comment by jodastephen** [comment]
Probably a static utility that converts each type to/from a matching JSR-310 type.

**Comment by raupachz** [comment]
I look into it if you don't mind. Need something to do in the Christmas season. Just let me know if you already have some pointers, otherwise I just try something and get back to you.

**Comment by jodastephen** [comment]
Great! I imagine something like

```
cass XmlDates() {
  static toOffsetDateTime(XmlGregorianCalendar)
  static toXmlGregorianCalendar(OffsetDateTime)
  static toDuration(java.xml.Duration)
  static toXmlDuration(java.time.Duration)
}
```

**Comment by raupachz** [comment]
Are you sure about `OffsetDateTime`? Problem is `XMLGregorianCalendar` allows undefined properties. Otherwise it would not be possible to map a `gDay` to `XMLGregorianCalendar` for example. It is however not possible to map `XMLGregorianCalendar` to `OffsetDateTime` if we miss the date and time.

Of course we could just throw exceptions if we encounter undefined properties, but this would make the converting between `XMLGregorianCalendar` and `OffsetDateTime` very limited.

Thoughts?

**Comment by jodastephen** [comment]
I've never had to actively use the class, so I don't know its model well. `OffsetDateTime` is conceptually the right class for XML dates and times based on the spec. Perhaps the conversion needs to default time fields to 0 and day/month to 1 if missing?

I note that the API of the class does have a `setTmiezone` method, but that is very much to do with offsets, not `ZoneId`.

Methods could be added to convert `LocalDate` to/from xml calendar, if that is a reasonable use case. Same with `LocalTime`. Really the class looks like a complex hash map, which isn't really very helpful.

**Comment by raupachz** [comment]
We could provide an adapter for `XMLGregorianCalendar` that implements `TemporalAccessor`" 

```
LocalDate ld = LocalDate.from(xmlGregorianCalendarAdapter);
YearMonth ym = YearMonth.from(xmlGregorianCalendarAdapter);
```

It is also possible to create a bunch of [XmlAdapter] (https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/annotation/adapters/XmlAdapter.html). This was previously done for [JAXB and Joda-Time](http://blog.bdoughan.com/2011/05/jaxb-and-joda-time-dates-and-times.html).

**Comment by jodastephen** [comment]
A `TemporalAccessor` wrapper would be a good one, although only in one direction!

Providing `XmlAdapter` would also make sense.

**Comment by raupachz** [comment]
[Who seeks shall find.](https://github.com/migesok/jaxb-java-time-adapters). That is the approach I was thinking about and is probably the way to go if you want to use JAXB with java.time. 

I am just not sure if this fits in threeten-extra. Seems more like a project of its own.

**Comment by raupachz** [comment]
Feeling a bit guilty because I still haven't provided a PR for this one. 

Just pushed a bunch of XmlAdapters for the classes in java.time and threeten-extra [on my fork](https://github.com/raupachz/threeten-extra/tree/jaxb).

```
@XmlRootElement
public class Dog implements Serializable {
    
    private String name;
    private LocalDate dayOfBirth;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @XmlJavaTypeAdapter(LocalDateAdapter.class)
    public LocalDate getDayOfBirth() {
        return dayOfBirth;
    }

    public void setDayOfBirth(LocalDate dayOfBirth) {
        this.dayOfBirth = dayOfBirth;
    }

}
```

Haven't used JAXB (and anything SOAP-like) in a long time so I am not sure if someone would use the code I provided. It does work on my pet projects. The implementation uses the `parse` and `toString` methods that nearly all classes in java.time and threeten-extra have.

Unit tests and javadoc still missing though.

Really looking forward to JSON-B and what it can provide but I wanted to finish this one before starting something new.

**Comment by mwkroening** [comment]
I am the current maintainer of [jaxb-java-time-adapters](https://github.com/jaxb-java-time-adapters/jaxb-java-time-adapters).
Would it make sense to integrate the JAXB adapters into this library?

**Comment by jodastephen** [comment]
I think its probably best to keep them separate as they have different dependencies. But there should be a "related projects" page on the ThreeTen-Extra website that links to those adapters.

**Comment by jodastephen** [comment]
https://github.com/ThreeTen/threeten-extra/commit/e1c5e79ac96d95875e01f1cedc1f7aa74f35e1e3

**Comment by mwkroening** [comment]
We moved the project from [migesok/jaxb-java-time-adapters](https://github.com/migesok/jaxb-java-time-adapters) to [jaxb-java-time-adapters/jaxb-java-time-adapters](https://github.com/jaxb-java-time-adapters/jaxb-java-time-adapters) to have a project group instead of having one user own the repo.

I have not jet released a new version under the new name.

Would you be okay with me renaming it to ThreeTen-JAXB/threeten-jaxb (and potentially ThreeTen-JAXB/threeten-extra-jaxb). I think that would be a much better name for the project.

**Comment by jodastephen** [comment]
I don't have a problem with threeten-jaxb, although I'd prefer it to be in lower case to keep a slight separation between the projects.

**Comment by mwkroening** [comment]
Thank you! I renamed it accordingly: [threeten-jaxb/threeten-jaxb](https://github.com/threeten-jaxb/threeten-jaxb)

**Comment by jodastephen** [comment]
I'm going to close this issue as it is unwieldy and unclear what needs doing if anything. If anyone wants to raise a new issue with a specific task, feel free. Otherwise we should assume that [threeten-jaxb/threeten-jaxb](https://github.com/threeten-jaxb/threeten-jaxb) is sufficient.
